proposals/lip-0053.md

```diff
@@ -252,234 +389,264 @@ crossChainUpdateTransactionParams = {
+        verifyCertificate(trs)
+    if len(trs.params.activeValidatorsUpdate) > 0 or trs.params.certificateThreshold != validators(trs.params.sendingChainID).certificateThreshold:
+        verifyValidatorsUpdate(trs)
+    if trs.params.inboxUpdate is not empty: # An empty object has all properties set to their default values.
+        verifyPartnerChainOutboxRoot(trs)
+```
 
-Furthermore, if the CCU also contains a non-empty `inboxUpdate`, the certificate is only valid if it allows the sidechain account to remain live for a reasonable amount of time. This is done by checking that:
+##### Execution
 
 ```python
-timestamp - CCU.params.certificate.timestamp < LIVENESS_LIMIT / 2
+def execute(trs: Transaction) -> None:
+    # Verify certificate signature. We do it here because if it fails, the transaction fails rather than being invalid.
+    verifyCertificateSignature(trs)
+    # Process cross-chain messages in inbox update.
+    crossChainMessages = trs.params.inboxUpdate.crossChainMessages
+
+    # First process basic checks for all CCMs.
+    for ccmBytes in crossChainMessages:
+        # Verify general format. Past this point, we can access ccm root properties.
+        try:
+            ccm = decode(crossChainMessageSchema, ccmBytes)
+            validateFormat(ccm)
+            # The CCM must come from the sending chain.
+            if trs.params.sendingChainID != ccm.sendingChainID:
+                raise Exception("CCM is not from the sending chain.")
+            if ccm.status == ERROR_CODE_CHANNEL_UNAVAILABLE:
+                raise Exception("CCM status channel unavailable can only be set on the mainchain.")
+        except:
+            terminateChain(trs.params.sendingChainID)
+            ccmID = sha256(ccmBytes)
+            emitEvent(
+                module = MODULE_NAME_INTEROPERABILITY,
+                name = EVENT_NAME_CCM_PROCESSED,
+                data = {"ccmID": ccmID, "result": ERROR_CODE_INVALID_FORMAT},
+                topics = [trs.params.sendingChainID, OWN_CHAIN_ID]
+            )
+            # In this case, we do not even update the chain account with the new certificate.
+            return
+
+    for ccmBytes in crossChainMessages:
+        appendToInboxTree(trs.params.sendingChainID, ccmBytes)
+        ccm = decode(crossChainMessageSchema, ccmBytes)
+
+        # If the receiving chain is the mainchain, apply the CCM.
+        if ccm.receivingChainID == CHAIN_ID_MAINCHAIN:
+            apply(trs, ccm)
+        # If the receiving chain is not the mainchain, forward the CCM.
+        elif ccm.receivingChainID != CHAIN_ID_MAINCHAIN:
+            forward(trs, ccm)
+    
+    # Update sidechain validators.
+    if len(trs.params.activeValidatorsUpdate) > 0 or trs.params.certificateThreshold != validators(trs.params.sendingChainID).certificateThreshold:
+        updateValidators(trs)
+    
+    # Update last certificate.
+    if len(trs.params.certificate) > 0:
+        updateLastCertificate(trs)
```

Comment:
This function is undefined (same in the sidechain CCU execute), did you mean `updateCertificate` ?

Reply:
yes, thanks